lib/mlxdevm.c

```diff
@@ -183,6 +183,67 @@ mlxdevm_sf_port_add(struct mlxdevm *dl, uint32_t pfnum, uint32_t sfnum)
 	return NULL;
 }
 
+static int cmd_port_dump_cb_to_list(const struct nlmsghdr *nlh, void *data)
+{
+	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
+	struct nlattr *tb[MLXDEVM_ATTR_MAX + 1] = {};
+	struct port_list_head *head = data;
+	struct mlxdevm_port_list *cur;
+	struct mlxdevm_port *port;
+
+	mnl_attr_parse(nlh, sizeof(*genl), attr_cb, tb);
+	if (!tb[MLXDEVM_ATTR_DEV_BUS_NAME] || !tb[MLXDEVM_ATTR_DEV_NAME] ||
+	    !tb[MLXDEVM_ATTR_PORT_INDEX] || !tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER] ||
+	    !tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER])
+		return MNL_CB_ERROR;
+
+	cur = calloc(1, sizeof(struct mlxdevm_port_list));
+	TAILQ_INSERT_TAIL(head, cur, entry);
+	port = &cur->port;
+
+	port->port_index = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_INDEX]);
+	port->ndev_ifindex = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_NETDEV_IFINDEX]);
+	port->pfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER]);
+	port->sfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER]);
+	cmd_port_fn_get(tb, port);
+	return MNL_CB_OK;
+}
+
+int mlxdevm_sf_port_dump_list(struct mlxdevm *dl, struct port_list_head *head)
+{
+	struct nlmsghdr *nlh;
+	int err;
+
+	nlh = mnlu_gen_socket_cmd_prepare(&dl->nlg, MLXDEVM_CMD_PORT_GET,
+					  NLM_F_REQUEST | NLM_F_ACK | NLM_F_DUMP);
+
+	mnl_attr_put_strz(nlh, MLXDEVM_ATTR_DEV_BUS_NAME, dl->bus);
+	mnl_attr_put_strz(nlh, MLXDEVM_ATTR_DEV_NAME, dl->dev);
+
+	err = mnlu_gen_socket_sndrcv(&dl->nlg, nlh, cmd_port_dump_cb_to_list, head);
+	if (err)
+		return err;
+
+	return 0;
+}
+
+void mlxdevm_sf_port_del_list_item(struct mlxdevm *dl, struct port_list_head *head,
+						struct mlxdevm_port_list *port)
+{
+	struct nlmsghdr *nlh;
+
+	nlh = mnlu_gen_socket_cmd_prepare(&dl->nlg, MLXDEVM_CMD_PORT_DEL,
```

Comment:
put the code of line 235 to 241 in a helper static function say

mlxdevm_port_del_cmd();

and call this function from here and also from mlxdevm_sf_port_del().

This is needed as both has duplicate code.